native/protocol/src/x3dh.rs

```diff
@@ -0,0 +1,364 @@
+    StaticSecret as X25519StaticSecret,
+};
+
+// See https://signal.org/docs/specifications/x3dh/ for an explanation of the below variable names
+// and functions of the X3DH protocol.
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKey {
+    pub pre_key: X25519PublicKey,
+    pub signature: Signature,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKeys {
+    pub pre_keys: Vec<X25519PublicKey>,
+    pub signature: Signature,
+}
+
+pub struct X3DHSendKeyAgreement {
+    pub ek: X25519PublicKey,
+    pub sk: [u8; 32],
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Message {
+    pub sender_identity: String,
+    pub sender_ik: VerifyingKey,
+    pub ek: X25519PublicKey,
+    pub opk: Option<X25519PublicKey>,
+    pub ciphertext: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PreKeyBundle {
+    pub ik: VerifyingKey,
+    pub opk: Option<X25519PublicKey>,
+    pub spk: SignedPreKey,
+}
+
+// KDF(KM) represents 32 bytes of output from the HKDF algorithm [3] with inputs:
+//    HKDF input key material = F || KM, where KM is an input byte sequence containing secret key material, and F is a byte sequence containing 32 0xFF bytes if curve is X25519, and 57 0xFF bytes if curve is X448. F is used for cryptographic domain separation with XEdDSA [2].
+//    HKDF salt = A zero-filled byte sequence with length equal to the hash output length.
+//    HKDF info = An ASCII string identifying the application.
+fn kdf(km: &[u8]) -> [u8; 32] {
+    let salt = [0; 32];
+    let f = [0xFF, 32];
+    let ikm = [&f, km].concat();
+    let hk = Hkdf::<Sha256>::new(Some(&salt), &ikm);
+    let mut okm = [0u8; 32];
+    hk.expand(b"Brongnal", &mut okm).unwrap();
+    okm
+}
+
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum X3DHError {
+    #[error("Signature failed to validate.")]
+    SignatureValidation,
+    #[error("Aead routine failed.")]
+    Aead(#[from] AeadError),
+}
+
+// If the bundle does not contain a one-time prekey, she calculates:
+//    DH1 = DH(IKA, SPKB)
+//    DH2 = DH(EKA, IKB)
+//    DH3 = DH(EKA, SPKB)
+//    SK = KDF(DH1 || DH2 || DH3)
+//If the bundle does contain a one-time prekey, the calculation is modified to include an additional DH:
+//    DH4 = DH(EKA, OPKB)
+//    SK = KDF(DH1 || DH2 || DH3 || DH4)
+fn initiate_send_get_sk(
+    recipient_ik: VerifyingKey,
+    spk: SignedPreKey,
+    opk: Option<X25519PublicKey>,
+    sender_ik: &SigningKey,
+) -> Result<X3DHSendKeyAgreement, X3DHError> {
+    let _ = verify_bundle(&recipient_ik, &[spk.pre_key], &spk.signature)
+        .map_err(|_| X3DHError::SignatureValidation);
+
+    let ek = X25519ReusableSecret::random();
+    let dh1 = X25519StaticSecret::from(sender_ik.to_scalar_bytes()).diffie_hellman(&spk.pre_key);
+    let dh2 = ek.diffie_hellman(&X25519PublicKey::from(
+        recipient_ik.to_montgomery().to_bytes(),
+    ));
+    let dh3 = ek.diffie_hellman(&spk.pre_key);
```

Comment:
[curiosity] Is this ratcheting? Give assigned reading, plzty

Reply:
This is not ratcheting. Ratcheting is [documented here](https://signal.org/docs/specifications/doubleratchet/). The derived shared secret is split. One half is used for encrypting the current message. Another is used in a hash chain to derive new shared secrets to provide better forward secrecy properties for subsequent messages.